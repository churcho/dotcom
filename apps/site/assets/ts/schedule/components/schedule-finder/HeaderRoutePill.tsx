import React, { ReactElement } from "react";
import isSilverLine, {
  isSilverLineWaterfront
} from "../../../helpers/silver-line";
import { RouteType } from "../../../__v3api";

interface Props {
  id: string;
  type: RouteType;
  name: string;
}

const HeaderRoutePill = ({
  id,
  type,
  name
}: Props): ReactElement<HTMLElement> | null =>
  type === 3 ? (
    <div className="m-route-pills">
      <div
        className={`h1 schedule-finder__modal-route-pill u-bg--${
          isSilverLine(id) ? "silver-line" : "bus"
        }`}
      >
        {isSilverLineWaterfront(id) ? "SL" : name}
      </div>
    </div>
  ) : null;

export default HeaderRoutePill;
