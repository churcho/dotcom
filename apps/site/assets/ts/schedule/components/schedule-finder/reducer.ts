import { SelectedDirection, SelectedOrigin } from "../__schedule";

export enum MODAL_ACTIONS {
  openModal,
  closeModal,
  setDirectionError,
  updateOriginSearch,
  selectDirection,
  selectOrigin,
  setErrors
}

// All the actions that can be performed on the schedule finder modal state
export type ModalAction =
  | { type: MODAL_ACTIONS.openModal; payload: string }
  | { type: MODAL_ACTIONS.closeModal; payload: string }
  | { type: MODAL_ACTIONS.updateOriginSearch; payload: string }
  | { type: MODAL_ACTIONS.selectDirection; payload: SelectedDirection }
  | { type: MODAL_ACTIONS.selectOrigin; payload: SelectedOrigin }
  | {
      type: MODAL_ACTIONS.setErrors;
      payload: { directionError?: boolean; originError?: boolean };
    };

// The state object of the schedule finder modals. Identical to prior
// ScheduleFinder component state, with an extra property to indicate whether
// the component is invoked from inside a modal
export interface ModalState {
  directionError: boolean;
  modalId: string | null;
  modalOpen: boolean;
  originError: boolean;
  originSearch: string;
  selectedDirection: SelectedDirection;
  selectedOrigin: SelectedOrigin;
  insideModal: boolean;
}

// Reducer function returning a new state object to reflect changes prompted by
// the specified action type and associated payload
export const scheduleFinderReducer = (
  previousState: ModalState,
  action: ModalAction
): ModalState => {
  switch (action.type) {
    // Open the modal and assign modalId
    case MODAL_ACTIONS.openModal: {
      return {
        ...previousState,
        modalId: action.payload,
        modalOpen: true
      };
    }
    // Handle the origin search user input
    case MODAL_ACTIONS.updateOriginSearch: {
      return {
        ...previousState,
        originSearch: action.payload
      };
    }
    // Assign new selected direction and reset origin
    case MODAL_ACTIONS.selectDirection: {
      if (previousState.selectedDirection !== action.payload) {
        return {
          ...previousState,
          selectedDirection: action.payload,
          selectedOrigin: null,
          directionError: action.payload === null,
          originError: false,
          originSearch: ""
        };
      }

      return previousState;
    }
    // Assign new selected origin
    case MODAL_ACTIONS.selectOrigin: {
      if (action.payload === null) {
        return {
          ...previousState,
          selectedOrigin: action.payload,
          modalOpen: previousState.modalId === "origin",
          originError: true
        };
      }
      // Selecting origin stop should open schedule modal
      return {
        ...previousState,
        selectedOrigin: action.payload,
        modalOpen: previousState.modalId === "origin",
        modalId: "schedule-sf",
        originError: false
      };
    }
    // Assign error state(s) if present
    case MODAL_ACTIONS.setErrors: {
      if (action.payload) {
        return {
          ...previousState,
          ...action.payload
        };
      }
      return previousState;
    }
    // Close modal
    case MODAL_ACTIONS.closeModal: {
      // Closing origin modal should open schedule modal
      if (action.payload === "origin") {
        return {
          ...previousState,
          modalId: previousState.selectedOrigin ? "schedule-sf" : null,
          modalOpen: !!previousState.selectedOrigin
        };
      }
      return {
        ...previousState,
        modalId: null,
        modalOpen: false
      };
    }
    // Otherwise don't change state
    default: {
      return previousState;
    }
  }
};
