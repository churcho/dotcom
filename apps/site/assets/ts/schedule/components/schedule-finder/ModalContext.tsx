import React, {
  createContext,
  useReducer,
  ReactElement,
  useContext
} from "react";
import { SelectedDirection, SelectedOrigin } from "../__schedule";
import { ModalAction, ModalState, scheduleFinderReducer } from "./reducer";

// The context value is an object with state and dispatch keys, populated here
// with initial values
const ModalContext = createContext<{
  state: ModalState;
  dispatch: (action: ModalAction) => void;
}>({
  // eslint-disable-next-line @typescript-eslint/no-object-literal-type-assertion
  state: {} as ModalState,
  dispatch: () => {}
});

// This component wraps ModalContext.Provider while also managing the context
// state, thereby enabling interaction between any context consumer components
// and the context state
const ModalProvider = ({
  selectedDirection,
  selectedOrigin,
  modalId,
  children,
  insideModal = false
}: {
  selectedDirection: SelectedDirection;
  selectedOrigin: SelectedOrigin;
  modalId: string | null;
  children: JSX.Element;
  insideModal?: boolean;
}): ReactElement => {
  const initialState: ModalState = {
    directionError: false,
    originError: false,
    originSearch: "",
    modalOpen: false,
    modalId: null,
    selectedDirection: null,
    selectedOrigin: null,
    insideModal: false
  };
  const [state, dispatch] = useReducer(scheduleFinderReducer, {
    ...initialState,
    selectedDirection,
    selectedOrigin,
    modalId,
    insideModal
  });

  return (
    <ModalContext.Provider value={{ state, dispatch }}>
      {children}
    </ModalContext.Provider>
  );
};

// A hook for reading and interacting with the context state. Can be used in any
// component with a <ModalProvider> ancestor, will expose the current state and
// a dispatch function that will handle actions via scheduleFinderReducer()
function useModalContext(): {
  state: ModalState;
  dispatch: (action: ModalAction) => void;
} {
  const { state, dispatch } = useContext(ModalContext);
  if (state === undefined || dispatch === undefined) {
    throw new Error("useModalContext must be used within a ModalProvider");
  }
  return { state, dispatch };
}

export { ModalContext, ModalProvider, useModalContext };
