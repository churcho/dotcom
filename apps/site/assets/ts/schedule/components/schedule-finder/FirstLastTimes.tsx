import React, { ReactElement } from "react";
import { SelectedDirection } from "../ScheduleFinder";
import { HoursOfOperation } from "../__schedule";

const getTimes = (
  hoursOfOperation: HoursOfOperation,
  serviceType: "week" | "saturday" | "sunday",
  selectedDirection: SelectedDirection
): [string, string] => [
  hoursOfOperation[serviceType][selectedDirection!].first_departure,
  hoursOfOperation[serviceType][selectedDirection!].last_departure
];

interface Props {
  hoursOfOperation: HoursOfOperation;
  selectedDirection: SelectedDirection;
}

const FirstLastTimes = ({
  hoursOfOperation,
  selectedDirection
}: Props): ReactElement<HTMLElement> | null => {
  if (selectedDirection === null) {
    return null;
  }

  const timesList = [
    [
      "Monday - Saturday",
      ...getTimes(hoursOfOperation, "week", selectedDirection)
    ],
    ["Sunday", ...getTimes(hoursOfOperation, "sunday", selectedDirection)]
  ];

  return (
    <div className="">
      {timesList.map(([label, firstTime, lastTime]) => (
        <>
          <h4>{label}</h4>
          <div className="first-last-trips">
            <div className="first-last-trips__trip">
              <span className="schedule-finder__first-last-trip-label">
                First Trip
              </span>
              {firstTime}
            </div>
            <div className="first-last-trips__trip">
              <span className="schedule-finder__first-last-trip-label">
                Last Trip
              </span>
              {lastTime}
            </div>
          </div>
        </>
      ))}
    </div>
  );
};

export default FirstLastTimes;
