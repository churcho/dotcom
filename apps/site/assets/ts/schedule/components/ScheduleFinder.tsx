import React, { ReactElement } from "react";
import { Route, DirectionId } from "../../__v3api";
import {
  SimpleStop,
  SimpleStopMap,
  RoutePatternsByDirection,
  ServiceInSelector,
  ScheduleNote as ScheduleNoteType,
  SelectedDirection,
  SelectedOrigin
} from "./__schedule";
import { handleReactEnterKeyPress } from "../../helpers/keyboard-events";
import icon from "../../../static/images/icon-schedule-finder.svg";
import renderSvg from "../../helpers/render-svg";
import Modal from "../../components/Modal";
import SelectContainer from "./schedule-finder/SelectContainer";
import ErrorMessage from "./schedule-finder/ErrorMessage";
import OriginModalContent from "./schedule-finder/OriginModalContent";
import ScheduleModalContent from "./schedule-finder/ScheduleModalContent";
import { ModalProvider, useModalContext } from "./schedule-finder/ModalContext";
import { MODAL_ACTIONS } from "./schedule-finder/reducer";

interface Props {
  services: ServiceInSelector[];
  route: Route;
  stops: SimpleStopMap;
  routePatternsByDirection: RoutePatternsByDirection;
  today: string;
  scheduleNote: ScheduleNoteType | null;
}

const parseSelectedDirection = (value: string): SelectedDirection => {
  if (value === "0") return 0;
  return 1;
};

const ScheduleFinder = ({
  route,
  services,
  stops,
  routePatternsByDirection,
  today,
  scheduleNote
}: Props): ReactElement<HTMLElement> => {
  const {
    direction_destinations: directionDestinations,
    direction_names: directionNames
  } = route;

  const validDirections = ([0, 1] as DirectionId[]).filter(
    direction => directionNames[direction] !== null
  );

  const { state, dispatch } = useModalContext();

  const handleUpdateOriginSearch = (searchQuery: string): void => {
    if (state.originSearch !== searchQuery) {
      dispatch({
        type: MODAL_ACTIONS.updateOriginSearch,
        payload: searchQuery
      });
    }
  };

  const handleSubmitForm = (): void => {
    if (state.selectedDirection === null || state.selectedOrigin === null) {
      dispatch({
        type: MODAL_ACTIONS.setErrors,
        payload: {
          directionError: state.selectedDirection === null,
          originError: state.selectedOrigin === null
        }
      });
      return; // don't open schedules modal
    }

    dispatch({ type: MODAL_ACTIONS.openModal, payload: "schedule-sf" });
  };

  const handleChangeOrigin = (origin: SelectedOrigin): void => {
    dispatch({ type: MODAL_ACTIONS.selectOrigin, payload: origin });
  };

  const handleOriginSelectClick = (): void => {
    if (state.selectedDirection === null) {
      dispatch({
        type: MODAL_ACTIONS.setErrors,
        payload: {
          directionError: state.selectedDirection === null
        }
      });
      return; // don't open origin modal
    }

    dispatch({ type: MODAL_ACTIONS.openModal, payload: "origin" });
  };

  return (
    <div className="schedule-finder">
      <h2 className="h3 schedule-finder__heading">
        {renderSvg("c-svg__icon", icon)} Schedule Finder
      </h2>
      <ErrorMessage
        directionError={state.directionError}
        originError={state.originError}
      />
      {!state.insideModal && (
        <div className="schedule-finder__helptext">
          Choose a stop to get schedule information and real-time departure
          predictions.
        </div>
      )}
      <div className="schedule-finder__modal-inputs">
        <div>
          <label
            className="schedule-finder__label"
            htmlFor="sf_direction_select"
          >
            Choose a direction
          </label>
          <SelectContainer
            error={state.directionError}
            id="sf_direction_select_container"
          >
            <select
              id="sf_direction_select"
              className="c-select-custom"
              value={
                state.selectedDirection !== null ? state.selectedDirection : ""
              }
              onChange={e => {
                dispatch({
                  type: MODAL_ACTIONS.selectDirection,
                  payload:
                    e.target.value === ""
                      ? null
                      : parseSelectedDirection(e.target.value)
                });
              }}
              onKeyUp={e =>
                handleReactEnterKeyPress(e, () => {
                  handleSubmitForm();
                })
              }
            >
              {!state.insideModal && <option value="">Select</option>}
              {validDirections.map(direction => (
                <option key={direction} value={direction}>
                  {directionNames[direction]!.toUpperCase()}{" "}
                  {directionDestinations[direction]!}
                </option>
              ))}
            </select>
          </SelectContainer>
        </div>
        <div>
          <label className="schedule-finder__label" htmlFor="sf_origin_select">
            Choose an origin stop
          </label>
          <SelectContainer
            error={state.originError}
            handleClick={handleOriginSelectClick}
            id="sf_origin_select_container"
          >
            <select
              id="sf_origin_select"
              className="c-select-custom c-select-custom--noclick"
              value={state.selectedOrigin || ""}
              onChange={e =>
                handleChangeOrigin(e.target.value ? e.target.value : null)
              }
              onKeyUp={e =>
                handleReactEnterKeyPress(e, () => {
                  handleSubmitForm();
                })
              }
            >
              <option value="">Select</option>
              {state.selectedDirection !== null
                ? stops[state.selectedDirection].map(
                    ({ id, name }: SimpleStop) => (
                      <option key={id} value={id}>
                        {name}
                      </option>
                    )
                  )
                : null}
            </select>
          </SelectContainer>
        </div>
      </div>

      {(!state.insideModal ||
        (state.insideModal && state.modalId === "origin")) && (
        <Modal
          openState={state.modalOpen}
          focusElementId={
            state.modalId === "origin" ? "origin-filter" : "modal-close"
          }
          ariaLabel={{
            label:
              state.modalId === "origin"
                ? "Choose Origin Stop"
                : "Choose Schedule"
          }}
          className={
            state.modalId === "origin" ? "schedule-finder__origin-modal" : ""
          }
          closeModal={() => {
            dispatch({
              type: MODAL_ACTIONS.closeModal,
              payload: state.modalId === "origin" ? "origin" : ""
            });
          }}
        >
          {() => (
            <>
              {state.modalId === "origin" && (
                <OriginModalContent
                  selectedDirection={state.selectedDirection}
                  selectedOrigin={state.selectedOrigin}
                  originSearch={state.originSearch}
                  stops={stops[state.selectedDirection!] || []}
                  handleChangeOrigin={handleChangeOrigin}
                  handleUpdateOriginSearch={handleUpdateOriginSearch}
                  directionId={state.selectedDirection!}
                />
              )}
              {state.modalId !== "origin" && (
                <ModalProvider
                  insideModal
                  modalId="schedule-sf"
                  selectedDirection={state.selectedDirection}
                  selectedOrigin={state.selectedOrigin}
                >
                  <ScheduleModalContent
                    route={route}
                    services={services}
                    stops={stops}
                    routePatternsByDirection={routePatternsByDirection}
                    today={today}
                    scheduleNote={scheduleNote}
                  />
                </ModalProvider>
              )}
            </>
          )}
        </Modal>
      )}

      {!state.insideModal && (
        <div className="schedule-finder__button text-right">
          <input
            className="btn btn-primary"
            type="submit"
            value="Get schedules"
            onClick={handleSubmitForm}
          />
        </div>
      )}
    </div>
  );
};

export default ScheduleFinder;
