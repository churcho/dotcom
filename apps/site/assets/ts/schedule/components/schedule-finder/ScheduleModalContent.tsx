import React, { ReactElement, useReducer, useEffect } from "react";
import UpcomingDepartures from "./UpcomingDepartures";
import { Route } from "../../../__v3api";
import {
  StopPrediction,
  RoutePatternsByDirection,
  ServiceInSelector,
  ScheduleNote as ScheduleNoteType,
  SelectedDirection,
  SelectedOrigin,
  UserInput,
  SimpleStopMap
} from "../__schedule";
import { reducer } from "../../../helpers/fetch";
import ServiceSelector from "./ServiceSelector";
import ScheduleNote from "../ScheduleNote";
import ScheduleFinder from "../ScheduleFinder";
import {
  ModalProvider,
  useModalContext
} from "./../schedule-finder/ModalContext";
import { MODAL_ACTIONS } from "./../schedule-finder/reducer";

type fetchAction =
  | { type: "FETCH_COMPLETE"; payload: StopPrediction[] }
  | { type: "FETCH_ERROR" }
  | { type: "FETCH_STARTED" };

export const fetchData = (
  routeId: string,
  selectedOrigin: SelectedOrigin,
  selectedDirection: SelectedDirection,
  dispatch: (action: fetchAction) => void
): Promise<void> => {
  dispatch({ type: "FETCH_STARTED" });
  return (
    window.fetch &&
    window
      .fetch(
        `/schedules/finder_api/departures?id=${routeId}&stop=${selectedOrigin}&direction=${selectedDirection}`
      )
      .then(response => {
        if (response.ok) return response.json();
        throw new Error(response.statusText);
      })
      .then(json => dispatch({ type: "FETCH_COMPLETE", payload: json }))
      // @ts-ignore
      .catch(() => dispatch({ type: "FETCH_ERROR" }))
  );
};

interface Props {
  route: Route;
  services: ServiceInSelector[];
  stops: SimpleStopMap;
  routePatternsByDirection: RoutePatternsByDirection;
  today: string;
  scheduleNote: ScheduleNoteType | null;
}

const ScheduleModalContent = ({
  route,
  services,
  stops,
  routePatternsByDirection,
  today,
  scheduleNote
}: Props): ReactElement<HTMLElement> | null => {
  const { id: routeId } = route;
  const [state, dispatch] = useReducer(reducer, {
    data: null,
    isLoading: true,
    error: false
  });
  const { state: modalState } = useModalContext();
  const { selectedDirection, selectedOrigin } = modalState;
  const hasOriginAndDirection =
    selectedOrigin !== null && selectedDirection !== null;

  useEffect(
    () => {
      if (hasOriginAndDirection) {
        fetchData(routeId, selectedOrigin, selectedDirection, dispatch);
      }
    },
    [routeId, selectedDirection, selectedOrigin]
  );

  const input: UserInput = {
    route: routeId,
    origin: selectedOrigin!,
    date: today,
    direction: selectedDirection
  };

  return (
    <>
      <ScheduleFinder
        route={route} // don't show for subway
        services={services}
        stops={stops}
        routePatternsByDirection={routePatternsByDirection}
        today={today}
        scheduleNote={scheduleNote}
      />
      {hasOriginAndDirection ? (
        <>
          <UpcomingDepartures state={state} input={input} />
          {scheduleNote ? (
            <ScheduleNote
              className="m-schedule-page__schedule-notes--modal"
              scheduleNote={scheduleNote}
            />
          ) : (
            <ServiceSelector
              services={services}
              routeId={routeId}
              routePatterns={routePatternsByDirection[selectedDirection!]}
              today={today}
            />
          )}
        </>
      ) : null}
    </>
  );
};

export default ScheduleModalContent;
