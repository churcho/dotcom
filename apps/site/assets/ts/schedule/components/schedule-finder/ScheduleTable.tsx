import React, { ReactElement } from "react";
import { EnhancedRoutePattern } from "../__schedule";
import { Journey } from "../__trips";
import TableRow from "./TableRow";
import { UserInput } from "../ScheduleFinder";

interface Props {
  journeys: Journey[];
  routePatterns: EnhancedRoutePattern[];
  input: UserInput;
}

const isSchoolTrip = (
  routePatternsById: {
    [key: string]: EnhancedRoutePattern;
  },
  routePatternId: string
): boolean =>
  (
    (routePatternsById[routePatternId] &&
      routePatternsById[routePatternId].time_desc) ||
    ""
  ).match(/school/gi) !== null;

const ScheduleTable = ({
  journeys,
  routePatterns,
  input
}: Props): ReactElement<HTMLElement> => {
  const routePatternsById = routePatterns.reduce(
    (accumulator, routePattern) => ({
      ...accumulator,
      [routePattern.id]: routePattern
    }),
    {}
  ) as {
    [key: string]: EnhancedRoutePattern;
  };

  if (journeys.length === 0) {
    return (
      <div className="callout schedule-table--empty">
        There is no scheduled service for this time period.
      </div>
    );
  }
  const firstTrip = journeys[0];
  const lastTrip = journeys.length > 1 ? journeys[journeys.length - 1] : null;

  const anySchoolTrips = Object.values(journeys).some(
    ({ trip: { route_pattern_id: routePatternId } }) =>
      isSchoolTrip(routePatternsById, routePatternId)
  );

  return (
    <>
      <div className="schedule-finder__first-last-trip--bordered first-last-trips">
        <div className="first-last-trips__trip">
          <span className="schedule-finder__first-last-trip-label">
            First Trip
          </span>
          {firstTrip.departure.time}
        </div>
        {lastTrip && (
          <div className="first-last-trips__trip">
            <span className="schedule-finder__first-last-trip-label">
              Last Trip
            </span>
            {lastTrip.departure.time}
          </div>
        )}
      </div>
      {anySchoolTrips && (
        <p>
          <strong>S</strong> - Does NOT run on school vacation
        </p>
      )}
      <table className="schedule-table">
        <thead className="schedule-table__header">
          <tr className="schedule-table__row-header">
            {anySchoolTrips && (
              <th
                scope="col"
                className="schedule-table__row-header-label--tiny"
              />
            )}
            <th scope="col" className="schedule-table__row-header-label">
              Departs
            </th>
            {firstTrip.route.type === 2 && (
              <th
                scope="col"
                className="schedule-table__row-header-label--small"
              >
                Train
              </th>
            )}
            <th scope="col" className="schedule-table__row-header-label">
              Destination
            </th>
            <th scope="col" className="schedule-table__th--flex-end">
              Trip Details
            </th>
          </tr>
        </thead>
        <tbody>
          {journeys.map((journey: Journey) => (
            <TableRow
              key={journey.trip.id}
              input={input}
              journey={journey}
              isSchoolTrip={isSchoolTrip(
                routePatternsById,
                journey.trip.route_pattern_id
              )}
              anySchoolTrips={anySchoolTrips}
            />
          ))}
        </tbody>
      </table>
    </>
  );
};

export default ScheduleTable;
