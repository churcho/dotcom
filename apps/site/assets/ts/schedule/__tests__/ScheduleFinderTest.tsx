import React from "react";
import { ReactWrapper, mount } from "enzyme";
import ScheduleFinder from "../components/ScheduleFinder";
import { EnhancedRoute } from "../../__v3api";
import {
  RoutePatternsByDirection,
  ServiceInSelector
} from "../components/__schedule";
import { ModalProvider } from "./../components/schedule-finder/ModalContext";
import WrappedModal from "../../components/Modal";

const services: ServiceInSelector[] = [
  {
    valid_days: [1, 2, 3, 4, 5],
    typicality: "typical_service",
    type: "weekday",
    start_date: "2019-07-08",
    removed_dates_notes: {},
    removed_dates: [],
    name: "Weekday",
    id: "BUS319-J-Wdy-02",
    end_date: "2019-08-30",
    description: "Weekday schedule",
    added_dates_notes: {},
    added_dates: [],
    rating_start_date: "2019-06-25",
    rating_end_date: "2019-10-25",
    rating_description: "Test",
    "default_service?": true
  },
  {
    valid_days: [6],
    typicality: "typical_service",
    type: "saturday",
    start_date: "2019-07-13",
    removed_dates_notes: {},
    removed_dates: [],
    name: "Saturday",
    id: "BUS319-K-Sa-02",
    end_date: "2019-08-31",
    description: "Saturday schedule",
    added_dates_notes: {},
    added_dates: [],
    rating_start_date: "2019-06-25",
    rating_end_date: "2019-10-25",
    rating_description: "Test",
    "default_service?": false
  },
  {
    valid_days: [7],
    typicality: "typical_service",
    type: "sunday",
    start_date: "2019-07-14",
    removed_dates_notes: {},
    removed_dates: [],
    name: "Sunday",
    id: "BUS319-L-Su-02",
    end_date: "2019-08-25",
    description: "Sunday schedule",
    added_dates_notes: {},
    added_dates: [],
    rating_start_date: "2019-06-25",
    rating_end_date: "2019-10-25",
    rating_description: "Test",
    "default_service?": false
  }
];

// the enzyme test was done as one test because there was
// an issue mounting it more than once due to the focus-trap
// dependency that the Modal component depends on

/* eslint-disable @typescript-eslint/camelcase */
const today = "2019-12-05";
const body = '<div id="react-root"></div>';
const route: EnhancedRoute = {
  alert_count: 0,
  description: "",
  direction_destinations: { 0: "Oak Grove", 1: "Forest Hills" },
  direction_names: { 0: "Inbound", 1: "Outbound" },
  header: "",
  id: "Orange",
  long_name: "Orange Line",
  name: "Orange",
  type: 1
};
const oneDirectionRoute: EnhancedRoute = {
  alert_count: 0,
  description: "",
  direction_destinations: { 0: "Destination", 1: null },
  direction_names: { 0: "Outbound", 1: null },
  header: "",
  id: "route",
  long_name: "the route",
  name: "Route",
  type: 1
};

const stops = {
  "1": [
    {
      name: "SL",
      id: "741",
      is_closed: false,
      zone: "1"
    },
    {
      name: "Abc",
      id: "123",
      is_closed: false,
      zone: null
    },
    {
      name: "Wellington",
      id: "place-welln",
      is_closed: true,
      zone: null
    }
  ],
  "0": [
    {
      name: "Wellington",
      id: "place-welln",
      is_closed: true,
      zone: null
    },
    {
      name: "Abc",
      id: "123",
      is_closed: false,
      zone: null
    },
    {
      name: "SL",
      id: "741",
      is_closed: false,
      zone: "1"
    }
  ]
};

const routePatternsByDirection = {
  "0": [
    {
      typicality: 1,
      time_desc: "School Trip",
      shape_id: "9840004",
      route_id: "CR-Fitchburg",
      representative_trip_id: "CR-Weekday-Spring-19-401",
      name: "North Station - Wachusett",
      headsign: "Wachusett",
      id: "CR-Fitchburg-0-0",
      direction_id: 0
    }
  ],
  "1": [
    {
      typicality: 1,
      time_desc: "School Trip",
      shape_id: "9840003",
      route_id: "CR-Fitchburg",
      representative_trip_id: "CR-Weekday-Spring-19-400",
      name: "Wachusett - North Station",
      headsign: "North Station",
      id: "CR-Fitchburg-0-1",
      direction_id: 1
    }
  ]
} as RoutePatternsByDirection;

it("renders", () => {
  const tree = mount(
    <ScheduleFinder
      route={route}
      stops={stops}
      services={services}
      routePatternsByDirection={routePatternsByDirection}
      today={today}
      scheduleNote={null}
    />,
    {
      wrappingComponent: ModalProvider,
      wrappingComponentProps: {
        modalId: "test",
        selectedDirection: 0,
        selectedOrigin: null
      }
    }
  );
  expect(tree.html()).toMatchSnapshot();
});

it("defaults to the sole direction for unidirectional routes", () => {
  const wrapper = mount(
    <ScheduleFinder
      route={oneDirectionRoute}
      stops={stops}
      services={services}
      routePatternsByDirection={routePatternsByDirection}
      today={today}
      scheduleNote={null}
    />,
    {
      wrappingComponent: ModalProvider,
      wrappingComponentProps: {
        modalId: "test",
        selectedDirection: 0,
        selectedOrigin: null
      }
    }
  );

  expect(wrapper.find("#sf_direction_select").prop("value")).toEqual(0);
});

describe("modal", () => {
  let wrapper: ReactWrapper;
  beforeEach(() => {
    wrapper = mount(
      <ScheduleFinder
        route={route}
        stops={stops}
        services={services}
        routePatternsByDirection={routePatternsByDirection}
        today={today}
        scheduleNote={null}
      />,
      {
        wrappingComponent: ModalProvider,
        wrappingComponentProps: {
          modalId: "test",
          selectedDirection: 0,
          selectedOrigin: null
        }
      }
    );
  });

  it("does not show when there are errors", () => {
    wrapper
      .find("#sf_direction_select")
      .simulate("change", { target: { value: "" } });

    wrapper
      .find("#sf_origin_select")
      .simulate("change", { target: { value: "" } });

    wrapper.find("input").simulate("click"); // "Get schedules"

    expect(wrapper.exists(".error-container")).toBeTruthy();

    // the schedule modal should not be showing
    expect(wrapper.find(WrappedModal).prop("openState")).toBeFalsy();
  });

  it("shows schedules when there are no errors", () => {
    wrapper
      .find("#sf_direction_select")
      .simulate("change", { target: { value: "0" } });

    wrapper
      .find("#sf_origin_select")
      .simulate("change", { target: { value: "place-welln" } });

    wrapper.find("input").simulate("click"); // "Get schedules"

    expect(wrapper.exists(".error-container")).toBeFalsy();

    // the schedule modal should be showing
    expect(wrapper.find(WrappedModal).prop("openState")).toBeTruthy();
  });

  it("can be controlled with keyboard input", () => {
    wrapper.find("#sf_origin_select").simulate("keyUp", { key: "Enter" });
    wrapper.find("#sf_direction_select").simulate("keyUp", { key: "Enter" });
    wrapper.find("#sf_origin_select").simulate("click");
  });

  it("shows origin picker only when there are no direction errors", () => {
    wrapper
      .find("#sf_direction_select")
      .simulate("change", { target: { value: "" } });

    wrapper
      .find("#sf_origin_select_container")
      .hostNodes()
      .simulate("click");

    expect(wrapper.exists(".error-container")).toBeTruthy();
    let originModal = wrapper
      .find(WrappedModal)
      .filterWhere(
        n => n.prop("className") === "schedule-finder__origin-modal"
      );

    expect(originModal.exists()).toBeFalsy();

    wrapper
      .find("#sf_direction_select")
      .simulate("change", { target: { value: "0" } });

    wrapper
      .find("#sf_origin_select_container")
      .hostNodes()
      .simulate("click");

    expect(wrapper.exists(".error-container")).toBeFalsy();

    originModal = wrapper
      .find(WrappedModal)
      .filterWhere(
        n => n.prop("className") === "schedule-finder__origin-modal"
      );

    expect(originModal.prop("openState")).toBeTruthy();
  });

  it("allows interactive search for origin stop", () => {
    wrapper
      .find("#sf_origin_select_container")
      .hostNodes()
      .simulate("click");

    expect(wrapper.find(".schedule-finder__origin-list-item").length).toBe(3);

    wrapper
      .find("input#origin-filter")
      .simulate("change", { target: { value: "Wellington" } });

    expect(wrapper.find(".schedule-finder__origin-list-item").length).toBe(1);

    wrapper
      .find("input#origin-filter")
      .simulate("change", { target: { value: "" } });

    // click origin modal line item
    wrapper
      .find(".schedule-finder__origin-list-item")
      .at(0)
      .simulate("click");

    // keyup on origin modal line item
    wrapper
      .find("#sf_origin_select_container")
      .hostNodes()
      .simulate("click");

    wrapper
      .find(".schedule-finder__origin-list-item")
      .at(2)
      .simulate("keyUp", { key: "Enter" });
  });

  afterEach(() => {
    wrapper.unmount();
  });
});
