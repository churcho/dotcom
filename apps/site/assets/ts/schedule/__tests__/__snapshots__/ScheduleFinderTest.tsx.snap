// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `"<div class=\\"schedule-finder\\"><h2 class=\\"h3 schedule-finder__heading\\"><span class=\\"notranslate c-svg__icon\\" aria-hidden=\\"true\\">SVG</span> Schedule Finder</h2><div class=\\"schedule-finder__helptext\\">Choose a stop to get schedule information and real-time departure predictions.</div><div class=\\"schedule-finder__modal-inputs\\"><div><label class=\\"schedule-finder__label\\" for=\\"sf_direction_select\\">Choose a direction</label><div id=\\"sf_direction_select_container\\" tabindex=\\"0\\" class=\\"c-select-custom__container \\" role=\\"button\\"><select id=\\"sf_direction_select\\" class=\\"c-select-custom\\"><option value=\\"\\">Select</option><option value=\\"0\\">INBOUND Oak Grove</option><option value=\\"1\\">OUTBOUND Forest Hills</option></select><span class=\\"notranslate c-svg__icon c-select-custom__arrow\\" aria-hidden=\\"true\\">SVG</span></div></div><div><label class=\\"schedule-finder__label\\" for=\\"sf_origin_select\\">Choose an origin stop</label><div id=\\"sf_origin_select_container\\" tabindex=\\"0\\" class=\\"c-select-custom__container \\" role=\\"button\\"><select id=\\"sf_origin_select\\" class=\\"c-select-custom c-select-custom--noclick\\"><option value=\\"\\">Select</option><option value=\\"place-welln\\">Wellington</option><option value=\\"123\\">Abc</option><option value=\\"741\\">SL</option></select><span class=\\"notranslate c-svg__icon c-select-custom__arrow\\" aria-hidden=\\"true\\">SVG</span></div></div></div><div class=\\"schedule-finder__button text-right\\"><input class=\\"btn btn-primary\\" type=\\"submit\\" value=\\"Get schedules\\"></div></div>"`;
